### PR TITLE
Allow non-wrapped tasks in disabled executors

### PR DESCRIPTION
**Pull request by mar-kolya**
Some executors cannot handle tasks that have been wrapped into
`{Runnable,Callable}Wrapper` because they require certain subclass of
`{Callable,Runnable}` in order to work. We have a test that
effectively disables instrumentation for such executors.

This change makes sure that tasks that do not need to be
wrapped (which essentially means anything that is not lambda) still
get traced in such executors. One notable example of affected executor
type is `ScheduledThreadPoolExecutor`.

### CODE REVIEW

File: dd-java-agent/instrumentation/java-concurrent/src/main/java/datadog/trace/instrumentation/java/concurrent/ExecutorInstrumentationUtils.java [hunk 1]
```diff
@@ -45,7 +50,12 @@ public static boolean shouldAttachStateToTask(final Object task, final Executor
     final State state = contextStore.putIfAbsent(task, State.FACTORY);
     final TraceScope.Continuation continuation = scope.capture();
     if (state.setContinuation(continuation)) {
-      log.debug("created continuation {} from scope {}, state: {}", continuation, scope, state);
+      log.debug(
+          "created continuation {} from scope {}, state: {}",
+          continuation,
+          scope,
+          state,
+          new RuntimeException());
```
[the remark this answers is not in the document]
Comment (mar-kolya) [reply dd-java-agent/instrumentation/java-concurrent/src/main/java/datadog/trace/instrumentation/java/concurrent/ExecutorInstrumentationUtils.java]: That was pushed by mistake, I'll remove that exception

File: dd-java-agent/instrumentation/java-concurrent/akka-2.5-testing/src/test/groovy/AkkaExecutorInstrumentationTest.groovy [hunk 2]
```diff
@@ -79,22 +66,21 @@ class AkkaExecutorInstrumentationTest extends AgentTestRunner {
...
-    new ForkJoinPool()                                                                            | akkaSubmitForkJoinTaskMethod
-    new ForkJoinPool()                                                                            | akkaInvokeForkJoinTaskMethod
-
-    new ThreadPoolExecutor(1, 1, 1000, TimeUnit.NANOSECONDS, new ArrayBlockingQueue<Runnable>(1)) | executeRunnableMethod
-    new ThreadPoolExecutor(1, 1, 1000, TimeUnit.NANOSECONDS, new ArrayBlockingQueue<Runnable>(1)) | submitRunnableMethod
-    new ThreadPoolExecutor(1, 1, 1000, TimeUnit.NANOSECONDS, new ArrayBlockingQueue<Runnable>(1)) | submitCallableMethod
+    poolImpl                                                                                      | name                   | method
+    new ThreadPoolExecutor(1, 1, 1000, TimeUnit.NANOSECONDS, new ArrayBlockingQueue<Runnable>(1)) | "execute Runnable"     | executeRunnable
+    new ThreadPoolExecutor(1, 1, 1000, TimeUnit.NANOSECONDS, new ArrayBlockingQueue<Runnable>(1)) | "submit Runnable"      | submitRunnable
+    new ThreadPoolExecutor(1, 1, 1000, TimeUnit.NANOSECONDS, new ArrayBlockingQueue<Runnable>(1)) | "submit Callable"      | submitCallable
```
[the remark this answers is not in the document]
Comment (mar-kolya) [reply dd-java-agent/instrumentation/java-concurrent/akka-2.5-testing/src/test/groovy/AkkaExecutorInstrumentationTest.groovy]: Fixed

File: dd-java-agent/instrumentation/java-concurrent/src/main/java/datadog/trace/instrumentation/java/concurrent/ExecutorInstrumentationUtils.java [hunk 3]
```diff
@@ -28,7 +30,10 @@ public static boolean shouldAttachStateToTask(final Object task, final Executor
     return (scope instanceof TraceScope
         && ((TraceScope) scope).isAsyncPropagating()
         && task != null
-        && !ExecutorInstrumentationUtils.isDisabled(executor));
+        // Disabled executors cannot accept wrapped tasks, non wrapped tasks (i.e. tasks with
+        // injected fields) should still work fine.
+        && !((task instanceof RunnableWrapper || task instanceof CallableWrapper)
+            && ExecutorInstrumentationUtils.isDisabledForWrappedTasks(executor)));
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/java-concurrent/src/main/java/datadog/trace/instrumentation/java/concurrent/ExecutorInstrumentationUtils.java]: Perhaps the instanceof check would be better inside the `isDisabledForWrappedTasks` method?

File: dd-trace-ot/src/main/java/datadog/opentracing/scopemanager/ContinuableScope.java [hunk 4]
```diff
@@ -159,7 +159,10 @@ public void close(final boolean closeContinuationScope) {
         if (closeContinuationScope) {
           ContinuableScope.this.close();
         } else {
-          openCount.decrementAndGet();
+          // Same as in 'close()' above.
+          if (openCount.decrementAndGet() == 0 && finishOnClose) {
+            spanUnderScope.finish();
```
Comment (tylerbenson) [inline dd-trace-ot/src/main/java/datadog/opentracing/scopemanager/ContinuableScope.java]: Why was this change needed? Is this fixing a bug?

File: dd-trace-ot/src/test/groovy/datadog/opentracing/scopemanager/ScopeManagerTest.groovy [hunk 5]
```diff
@@ -183,6 +183,78 @@ class ScopeManagerTest extends Specification {
...
+
+  def "Continuation.close doesn't close parent scope, span finishes"() {
+    /*
+    This is highly confusing behaviour. Sequence of events is as following:
+      * Scope gets created along with span and with finishOnClose == true.
+      * Continuation gets created for that scope.
+      * Scope is closed.
+        At this point scope is not really closed. It is removed from scope
+        stack, but it is still alive because there is a live continuation attached
+        to it. This also means span is not closed.
```
Comment (tylerbenson) [inline dd-trace-ot/src/test/groovy/datadog/opentracing/scopemanager/ScopeManagerTest.groovy]: This is odd.  I think there's some desire to ensure the span doesn't close before the child, but I don't recall exactly.

File: dd-java-agent/instrumentation/java-concurrent/src/main/java/datadog/trace/instrumentation/java/concurrent/ExecutorInstrumentationUtils.java (hunk 3 above)
Comment (mar-kolya) [reply dd-java-agent/instrumentation/java-concurrent/src/main/java/datadog/trace/instrumentation/java/concurrent/ExecutorInstrumentationUtils.java]: Updated

File: dd-trace-ot/src/main/java/datadog/opentracing/scopemanager/ContinuableScope.java (hunk 4 above)
Comment (mar-kolya) [reply dd-trace-ot/src/main/java/datadog/opentracing/scopemanager/ContinuableScope.java]: Yes, see `Continuation.close doesn't close parent scope, span finishes` test in `ScopeManagerTest.groovy`. It describes sequence of events that triggers this. Removing this would make that test and also some tests in java-concurrent fail.
